### PR TITLE
`warn-pr-from-master` - Restore feature

### PR DESCRIPTION
**Pull request by kovsu**
Closes: #7670 

## Test URLs

https://github.com/refined-github/refined-github/compare/sandbox/keep-branch...yakov116:upstream

## Screenshot

image

### CODE REVIEW

File: source/features/warn-pr-from-master.tsx [hunk 1]
```diff
@@ -38,3 +43,11 @@ void features.add(import.meta.url, {
 	deduplicate: 'has-rgh',
```
Comment (fregante) [inline source/features/warn-pr-from-master.tsx:43]: Both of these can be dropped.

Also, is the banner removed/hidden when the user selects another branch?

**Review by fregante — APPROVED** [review]

**Comment by kovsu** [comment]
https://github.com/user-attachments/assets/e0b6c0b7-f14f-46ed-b90a-04769960eaa5

File: source/features/warn-pr-from-master.tsx [hunk 2]
```diff
@@ -27,14 +28,24 @@ async function init(): Promise<false | void> {
...
-	awaitDomReady: true,
-	deduplicate: 'has-rgh',
 	init,
 });
+
+/*
+
+Test URLs
+
+https://github.com/refined-github/refined-github/compare/sandbox/keep-branch...yakov116:upstream
```
Comment (fregante) [inline source/features/warn-pr-from-master.tsx:49]: The `upstream` name is confusing, let's use this one

```suggestion
https://github.com/refined-github/refined-github/compare/main...fregante:refined-github:main
```

File: source/features/warn-pr-from-master.tsx [hunk 3]
```diff
@@ -27,14 +28,24 @@ async function init(): Promise<false | void> {
...
+	observe('.is-cross-repo', warnPR, {signal});
+}
+
 void features.add(import.meta.url, {
 	include: [
 		pageDetect.isCompare,
 	],
 	exclude: [
 		pageDetect.isBlank,
 	],
```
Comment (fregante) [inline source/features/warn-pr-from-master.tsx:41]: This can also be dropped, see https://github.com/refined-github/refined-github/issues/7000#issuecomment-2287762931

```suggestion
```

**Review by fregante — APPROVED** [review]

File: source/features/warn-pr-from-master.tsx [hunk 4]
```diff
@@ -1,15 +1,16 @@
...
 import {getRepo} from '../github-helpers/index.js';
+import observe from '../helpers/selector-observer.js';
 
-async function init(): Promise<false | void> {
+async function warnPR(crossRepo: Element): Promise<false | void> {
 	let defaultBranch;
-	if (elementExists('.is-cross-repo')) {
-		const forkedRepository = getRepo($('[title^="head: "]')!.textContent)!;
+	if (crossRepo) {
+		const forkedRepository = getRepo($('.range-cross-repo-pair:has(#head-ref-selector) span[data-menu-button]')!.textContent)!;
```
Comment (fregante) [inline source/features/warn-pr-from-master.tsx:13]: This needs further changes actually. This condition is now always true, so `else` is never called.

But I don’t think it ever made sense, it was added in https://github.com/refined-github/refined-github/pull/3511

So:

- drop the `if(crossRepo)` line
- drop the entire `else` part
- Move the `data-menu-button` selector to the observer

**Review by fregante — CHANGES_REQUESTED** [review]

File: source/features/warn-pr-from-master.tsx (hunk 4 above)
Comment (fregante) [reply source/features/warn-pr-from-master.tsx:13]: I'm surprised that neither TS nor ESLint is complaining that the condition is always truthy 😬

File: source/features/warn-pr-from-master.tsx [hunk 5]
```diff
@@ -1,15 +1,16 @@
...
+import {$} from 'select-dom';
 import * as pageDetect from 'github-url-detection';
 
 import features from '../feature-manager.js';
 import getDefaultBranch, {defaultBranchOfRepo} from '../github-helpers/get-default-branch.js';
 import {getRepo} from '../github-helpers/index.js';
+import observe from '../helpers/selector-observer.js';
 
-async function init(): Promise<false | void> {
+async function warnPR(crossRepo: Element): Promise<false | void> {
```
Comment (fregante) [inline source/features/warn-pr-from-master.tsx:10]: This this is no longer an `init` function, it doesn't need to return `false` anymore`, it can just `return;`

```suggestion
async function warnPR(crossRepo: Element): Promise<void> {
```

File: source/features/warn-pr-from-master.tsx (hunk 4 above)
Comment (fregante) [reply source/features/warn-pr-from-master.tsx:13]: This is weird actually, we don't need to read this data from the DOM, it's already in the URL, it's being checked just later.

So probably:

- keep `observe('.is-cross-repo'`
- read `forkedRepository` from the `...user:repo` URL part instead


Also make sure that renamed repos are supported, test URL:

- https://github.com/refined-github/sandbox/compare/default-a...bfred-it-org:github-sandbox:main?expand=1

**Comment by fregante** [comment]
> I think it reload the page. Is it necessary to use observe?

You're right, it isn't, but it allows us to call the API earlier. Maybe we can drop the observer (which isn't necessary since all the info we need is in the URL) and use elementReady() instead of $()

If it works as is, for both test URLs, it's fine to merge, but ideally the feature should be rewritten:

- extract a function to get the head branch/repo information from the URL 
- add tests
- perhaps move this custom condition to asLongAs instead of the init
- use the new element ready function based on the observer (I forget the name, but it's in the same observer source file

Basically:

```js
async function init() {
  const anchor = await elementReady('the position')
  anchor.before(our banner)
}

async function isCrossRepoCompareFromMaster() {
  const {base, head} = parseCompareUrl(location.pathname)
  return head.repo !== base.repo && head.branch === await getDefaultBranch(head.repo)
}

features.add({
  init,
  asLongAs: [
    isCompare,
    isCrossRepoCompareFromMaster,
  ],
})
```

**Review by fregante — APPROVED** [review]

**Comment by kovsu** [comment]
@fregante I will try it after this pr merge. I think first thing we need to do is to get the feature working.

**Review by fregante — CHANGES_REQUESTED** [review]
I think it broke in a recent commit. This is why I was suggesting to just parse the URL instead

Screenshot 1

**Comment by kovsu** [comment]
https://github.com/user-attachments/assets/6cf07101-d965-4c53-bf22-83ebdb32c66b

**Comment by kovsu** [comment]
I mess it up 😂. Open a new pr.
